pkg/metrics/prometheus_metrics.go

```diff
@@ -0,0 +1,84 @@
+package metrics
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/klog/klogr"
+)
+
+var (
+	logger       = klogr.New().WithName("prometheus_exporter")
+	metricLabels = []string{"namespace", "metric", "selector"}
+	scalerErrors = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "keda_hpa",
+			Subsystem: "scaler",
+			Name:      "errors",
+			Help:      "Number of scaler errors",
+		},
+		metricLabels,
+	)
+	scalerMetricsValue = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "keda_hpa",
+			Subsystem: "scaler",
+			Name:      "metrics_value",
+			Help:      "Metric Value",
+		},
+		metricLabels,
+	)
+	errorTotals = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "keda_hpa",
+			Subsystem: "scaler",
+			Name:      "error_totals",
+			Help:      "Total number of scaler errors",
+		},
+		[]string{},
+	)
+)
+
+type prometheusMetricsExporter struct {
+	registry *prometheus.Registry
+}
+
+// NewPrometheusMetricsExporter return a new instance of the prometheus metrics server
+func NewPrometheusMetricsExporter() Exporter {
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(scalerErrors)
+	registry.MustRegister(scalerMetricsValue)
+	registry.MustRegister(errorTotals)
```

Comment:
I'm not really familiar with these prometheus apis, but from the usage of those vars (`scalerErrors`, `scalerMetricValue`, `errorTotals`) below, it seems that they are stateful per-exporter, right?

It's not a big deal since `NewPrometheusMetricsExporter` is only called once, but I think the contract of a New function usually would not assume shared state between returned values.

Reply:
Thank you @ahmelsayed. The vars represent the metrics to be tracked (registered) with the Prometheus client registry which is used to expose the metrics for collection (the http endpoint). I suppose it would be better to move the registration of the metrics to an init function. I could also move the registry out of the factory method and just reference it directly when starting the Prometheus client handler since it is sharing state.